grpc-context-utils/src/main/java/org/hypertrace/core/grpcutils/context/jwt/JwtParser.java

```diff
@@ -0,0 +1,94 @@
+package org.hypertrace.core.grpcutils.context.jwt;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.auth0.jwt.interfaces.JWTVerifier;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+import org.hypertrace.core.grpcutils.context.RequestContextConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JwtParser {
+  private static final Logger LOG = LoggerFactory.getLogger(JwtParser.class);
+  private static final String BEARER_TOKEN_PREFIX = "Bearer ";
+
+  private final JwtVerifierFactory jwtVerifierFactory;
+  private final Cache<String, Optional<VerifiedJwt>> verifiedJwtCache =
+      CacheBuilder.newBuilder().maximumSize(1000).expireAfterAccess(1, TimeUnit.HOURS).build();
+
+  public JwtParser(String expectedJwtAudience, String jwtIssuer) {
+    this.jwtVerifierFactory = new JwtVerifierFactory(expectedJwtAudience, jwtIssuer);
+  }
+
+  public Optional<VerifiedJwt> fromRequestContext(RequestContext requestContext) {
+    return requestContext
+        .get(RequestContextConstants.AUTHORIZATION_HEADER)
+        .flatMap(this::fromAuthHeader);
+  }
+
+  public Optional<VerifiedJwt> fromAuthHeader(String authHeaderValue) {
+    if (authHeaderValue.startsWith(BEARER_TOKEN_PREFIX)) {
+      return this.fromJwt(authHeaderValue.substring(BEARER_TOKEN_PREFIX.length()));
+    }
+
+    return Optional.empty();
+  }
+
+  public Optional<VerifiedJwt> fromJwt(String jwtValue) {
+    try {
+      return this.verifiedJwtCache.get(jwtValue, () -> this.decodeAndVerify(jwtValue));
```

Comment:
why use a cache ?

Reply:
if we need to access it multiple times we could store in the context object instead ?

Reply:
So my understanding is that we actually don't need want to force verification here, instead relying on the proxy to do it. Either way though, everything around JWTs should be package protected, we only want to expose the abstracted information via the context

Reply:
Yes. I think thats what we decided the other day. Verification is the responsibility of the proxy. 
Should we just add both the APIs then? Depending on the use case, the callers can decide.

Reply:
What do you mean by both the apis? the verified and unverified one?

Reply:
Removed jwt verification and now, exposing the abstracted information via the context.

Reply:
Yes. I meant `decodeAndVerify` and just `decode`

Reply:
I think we can add the verification part later on if required(we may not even need that if we're relying on proxy for verification). Also, the verification part requires passing in some extra values which makes it messy (which we may be able to avoid).